deployment/src/main/java/com/github/mcollovati/quarkus/hilla/deployment/FilterableRepositoryImplementor.java

```diff
@@ -0,0 +1,84 @@
+
+import java.util.List;
+import java.util.function.BiFunction;
+
+import dev.hilla.crud.filter.Filter;
+import io.quarkus.deployment.util.JandexUtil;
+import io.quarkus.gizmo.ClassTransformer;
+import io.quarkus.gizmo.MethodCreator;
+import io.quarkus.gizmo.MethodDescriptor;
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.ClassType;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Type;
+import org.objectweb.asm.ClassVisitor;
+import org.springframework.data.domain.Pageable;
+
+import com.github.mcollovati.quarkus.hilla.crud.FilterableRepositorySupport;
+
+public class FilterableRepositoryImplementor implements BiFunction<String, ClassVisitor, ClassVisitor> {
+
+    private final IndexView index;
+    private final DotName filterableRepositoryInterface;
+
+    public FilterableRepositoryImplementor(IndexView index, DotName filterableRepositoryInterface) {
+        this.index = index;
+        this.filterableRepositoryInterface = filterableRepositoryInterface;
+    }
+
+    @Override
+    public ClassVisitor apply(String className, ClassVisitor classVisitor) {
+        ClassInfo repositoryInterface = index.getClassByName(className);
+        List<Type> types =
+                JandexUtil.resolveTypeParameters(repositoryInterface.name(), filterableRepositoryInterface, index);
+        if (!(types.get(0) instanceof ClassType)) {
+            throw new IllegalArgumentException(
+                    "Entity generic argument of " + className + " is not a regular class type");
+        }
+        DotName entityType = types.get(0).name();
+        ClassTransformer transformer = new ClassTransformer(className);
+
+        MethodCreator countCreator = transformer.addMethod("count", long.class, Filter.class);
+        countCreator.returnValue(countCreator.invokeStaticMethod(
+                MethodDescriptor.ofMethod(
+                        FilterableRepositorySupport.class.getName(),
+                        "count",
+                        long.class.getName(),
+                        Filter.class.getName(),
+                        Class.class.getName()),
+                countCreator.getMethodParam(0),
+                countCreator.loadClass(entityType.toString())));
+
+        MethodCreator listCreator = transformer.addMethod("list", List.class, Pageable.class, Filter.class);
+        listCreator.returnValue(listCreator.invokeStaticMethod(
+                MethodDescriptor.ofMethod(
+                        FilterableRepositorySupport.class.getName(),
+                        "list",
+                        List.class.getName(),
+                        Pageable.class.getName(),
+                        Filter.class.getName(),
+                        Class.class.getName()),
+                listCreator.getMethodParam(0),
+                listCreator.getMethodParam(1),
+                listCreator.loadClassFromTCCL(entityType.toString())));
```

Comment:
Whats the reason to handle it this way? Is the Pageable always present here?

Reply:
Sorry, I don't understand the question.
Can you elaborate a bit?

Reply:
Could you explain what the code does, maybe this will solve my question

Reply:
Given, for example, an interface or class `MyRepo<MyEntity, Long> implements/extends FilterableRepository<MyEntity, Long>`, the code is implementing at build time the `list(Pageable, Filter)` method by delegating the call to a support class with an additional parameter that is the JPA entity class.
The implementation resulting from the above instruction is basically

```java
List<T> list(Pageable pageable, Filter filter) {
    return FilterableRepositorySupport.list(pageable, filter, MyEntity.class);
}
```

Reply:
To further clarity, the principal reason is being able to capture the entity type from the repository interface declaration, that would not be possible at runtime because of type erasure

Reply:
Ah ok and thats necessary because the delegated method otherwise wouldn't know about it?

Reply:
Exactly